var/spack/repos/builtin/packages/amrex/package.py

```diff
@@ -0,0 +1,93 @@
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from spack import *
+
+
+class Amrex(CMakePackage):
+    """A Block-Structured AMR Framework. AMReX is the successor to BoxLib."""
+
+    homepage = "https://ccse.lbl.gov/AMReX/index.html"
+    url      = "https://github.com/AMReX-Codes/amrex.git"
+
+    version('develop', git='https://github.com/AMReX-Codes/amrex.git', tag='master')
+    version('17.06', git='https://github.com/AMReX-Codes/amrex.git', commit='836d3c7')
+
+    variant('mpi', default=True, description='Enable MPI parallel support')
+# variant('omp', default=False, description='Enable OpenMP parallel support')
+
+    variant('dims',
+        default='3',
+        values=('1', '2', '3'),
+        multi=False,
+        description='Number of spatial dimensions')
+
+    variant('prec',
+        default='DOUBLE',
+        values=('FLOAT', 'DOUBLE'),
+        multi=False,
+        description='Floating point precision')
+
+    variant('fortran', default=True, description='Enable Fortran support')
+    variant('debug', default=False, description='Enable debugging features')
+    variant('particles', default=False, description='Include particle classes in build')
+
+    depends_on('mpi', when='+mpi')
+
+    def cmake_args(self):
+        spec = self.spec
+
+        cmake_args = [
+            '-DENABLE_POSITION_INDEPENDENT_CODE=ON',
+            '-DBL_SPACEDIM:INT=%d' % int(spec.variants['dims'].value),
+            '-DBL_PRECISION:STRING=%s' % spec.variants['prec'].value,
+        ]
+
+        if '+fortran' in spec:
+            cmake_args += [
+                '-DENABLE_FMG=ON',
+                '-DENABLE_FBASELIB=ON'
+            ]
+
+        if '+debug' not in spec:
+            cmake_args += ['-DBL_DEBUG:INT=0']
+
+        if '+particles' in spec:
+            cmake_args += ['-DBL_USE_PARTICLES:INT=1']
+
+        if '~mpi' in spec:
+            cmake_args += ['-DENABLE_MPI:INT=0']
+        else:
+            cmake_args += [
+                '-DCMAKE_C_COMPILER=%s' % spec['mpi'].mpicc,
+                '-DCMAKE_CXX_COMPILER=%s' % spec['mpi'].mpicxx
+            ]
+            if '+fortran' in spec:
+                cmake_args += [
+                    '-DCMAKE_Fortran_COMPILER=%s' % spec['mpi'].mpifc
+                ]
+            else:
+                cmake_args += ['-DENABLE_FORTRAN_MPI=OFF']
```

Comment:
> I didn't think it was the preferred modality for Spack to muck with a package's build variables unless it really needed to to affect a desired outcome.

If you add a variant you're promising that `+foo` means "with foo" and `~foo` means "without whatever foo is".  If the variant exists, it *will* be set on the final spec for the package, which means it can be used to resolve dependencies that require either `+foo` or `~foo`.  So, if you rely on the build's default for a variant, you set yourself up to violate that in the future if the default changes.  e.g., suppose `foo` is by default disabled, but then becomes enabled in the build defaults.  If you *only* supply the argument when you want to explicitly enable `foo`, `pkg+foo` and `pkg~foo` will erroneously be the same thing when the build system changes.  I think the important thing to realize here is that `~foo` is *not* the same as not having the variant `foo` at all.  `~foo` guarantees a particular configuration just as much as `+foo` does, while an absent `foo` guarantees nothing.

This isn't a really common case, so it's a minor thing, but maybe that gives you a sense of how the logic works.

>Related to this, what are the standards for handling things like debug/optimized, fpic, fortan (and other language api bindings including variants of fortran), etc. It seems like a lot of packages need to deal with these sorts of things and without some set of standards offered by Spack, the duplicate logic for them, make different assumptions about defaults, etc.

We don't currently have this written down and haven't settled on anything, but multi-value variants have been merged since the last time the discussion came up, so maybe there is some hope.  I would grep around in the packages and see what other packages do for the time being.